File: services/account/server/src/test/java/base/account/test/spy/AccountRepositorySpy.java
package base.account.test.spy;

import base.account.model.Account;
import base.account.repository.AccountRepository;

public class AccountRepositorySpy implements AccountRepository {
    private Account[] readByIdAndNameReturnValue;
    private int readByIdAndNameInvokeCount;
    private String readByIdAndNameIdArgument;
    private String readByIdAndNameNameArgument;
    private Account createReturnValue;
    private int createInvokeCount;
    private Account createAccountArgument;
    private Account updateByIdAndNameReturnValue;
    private int updateByIdAndNameInvokeCount;
    private String updateByIdAndNameIdArgument;
    private String updateByIdAndNameNameArgument;
    private Account updateByIdAndNameAccountArgument;
    private int deleteByIdAndNameReturnValue;
    private int deleteByIdAndNameInvokeCount;
    private String deleteByIdAndNameIdArgument;
    private String deleteByIdAndNameNameArgument;

    @Override
    public Account[] readByIdAndName(String id, String name) {
        readByIdAndNameInvokeCount++;
        readByIdAndNameIdArgument = id;
        readByIdAndNameNameArgument = name;
        return readByIdAndNameReturnValue;
    }

    @Override
    public Account create(Account account) {
        createInvokeCount++;
        createAccountArgument = account;
        return createReturnValue;
    }

    @Override
    public Account updateByIdAndName(String id, String name, Account account) {
        updateByIdAndNameInvokeCount++;
        updateByIdAndNameIdArgument = id;
        updateByIdAndNameNameArgument = name;
        updateByIdAndNameAccountArgument = account;
        return updateByIdAndNameReturnValue;
    }

    @Override
    public int deleteByIdAndName(String id, String name) {
        deleteByIdAndNameInvokeCount++;
        deleteByIdAndNameIdArgument = id;
        deleteByIdAndNameNameArgument = name;
        return deleteByIdAndNameReturnValue;
    }

    public void resetSpy() {
        readByIdAndNameInvokeCount = 0;
        readByIdAndNameIdArgument = null;
        readByIdAndNameNameArgument = null;
        createInvokeCount = 0;
        createAccountArgument = null;
        updateByIdAndNameInvokeCount = 0;
        updateByIdAndNameIdArgument = null;
        updateByIdAndNameNameArgument = null;
        updateByIdAndNameAccountArgument = null;
        deleteByIdAndNameInvokeCount = 0;
        deleteByIdAndNameIdArgument = null;
        deleteByIdAndNameNameArgument = null;
    }

    public void setReadByIdAndNameReturnValue(Account[] readByIdAndNameReturnValue) {
        this.readByIdAndNameReturnValue = readByIdAndNameReturnValue;
    }

    public int getReadByIdAndNameInvokeCount() {
        return readByIdAndNameInvokeCount;
    }

    public String getReadByIdAndNameIdArgument() {
        return readByIdAndNameIdArgument;
    }

    public String getReadByIdAndNameNameArgument() {
        return readByIdAndNameNameArgument;
    }

    public void setCreateReturnValue(Account createReturnValue) {
        this.createReturnValue = createReturnValue;
    }

    public int getCreateInvokeCount() {
        return createInvokeCount;
    }

    public Account getCreateAccountArgument() {
        return createAccountArgument;
    }

    public void setUpdateByIdAndNameReturnValue(Account updateByIdAndNameReturnValue) {
        this.updateByIdAndNameReturnValue = updateByIdAndNameReturnValue;
    }

    public int getUpdateByIdAndNameInvokeCount() {
        return updateByIdAndNameInvokeCount;
    }

    public String getUpdateByIdAndNameIdArgument() {
        return updateByIdAndNameIdArgument;
    }

    public String getUpdateByIdAndNameNameArgument() {
        return updateByIdAndNameNameArgument;
    }

    public Account getUpdateByIdAndNameAccountArgument() {
        return updateByIdAndNameAccountArgument;
    }

    public void setDeleteByIdAndNameReturnValue(int deleteByIdAndNameReturnValue) {
        this.deleteByIdAndNameReturnValue = deleteByIdAndNameReturnValue;
    }

    public int getDeleteByIdAndNameInvokeCount() {
        return deleteByIdAndNameInvokeCount;
    }

    public String getDeleteByIdAndNameIdArgument() {
        return deleteByIdAndNameIdArgument;
    }

    public String getDeleteByIdAndNameNameArgument() {
        return deleteByIdAndNameNameArgument;
    }
}

File: services/account/src/test/spy/account-repository-spy.js
import { AccountRepository } from '../../main/repository/account-repository.js';

class AccountRepositorySpy extends AccountRepository {
  #readByNameReturnValue;
  #readByNameInvokeCount = 0;
  #readByNameNameArgument;
  #readByIdAndNameReturnValue;
  #readByIdAndNameInvokeCount = 0;
  #readByIdAndNameIdArgument;
  #readByIdAndNameNameArgument;
  #createReturnValue;
  #createInvokeCount = 0;
  #createAccountArgument;
  #updateByIdAndNameReturnValue;
  #updateByIdAndNameInvokeCount = 0;
  #updateByIdAndNameIdArgument;
  #updateByIdAndNameNameArgument;
  #updateByIdAndNameAccountArgument;
  #deleteByIdAndNameReturnValue;
  #deleteByIdAndNameInvokeCount = 0;
  #deleteByIdAndNameIdArgument;
  #deleteByIdAndNameNameArgument;

  async readByName(name) {
    this.#readByNameInvokeCount++;
    this.#readByNameNameArgument = name;
    return this.#readByNameReturnValue;
  }

  async readByIdAndName(id, name) {
    this.#readByIdAndNameInvokeCount++;
    this.#readByIdAndNameIdArgument = id;
    this.#readByIdAndNameNameArgument = name;
    return this.#readByIdAndNameReturnValue;
  }

  async create(account) {
    this.#createInvokeCount++;
    this.#createAccountArgument = account;
    return this.#createReturnValue;
  }

  async updateByIdAndName(id, name, account) {
    this.#updateByIdAndNameInvokeCount++;
    this.#updateByIdAndNameIdArgument = id;
    this.#updateByIdAndNameNameArgument = name;
    this.#updateByIdAndNameAccountArgument = account;
    return this.#updateByIdAndNameReturnValue;
  }

  async deleteByIdAndName(id, name) {
    this.#deleteByIdAndNameInvokeCount++;
    this.#deleteByIdAndNameIdArgument = id;
    this.#deleteByIdAndNameNameArgument = name;
    return this.#deleteByIdAndNameReturnValue;
  }

  resetSpy() {
    this.#readByNameInvokeCount = 0;
    this.#readByNameNameArgument = undefined;
    this.#readByIdAndNameInvokeCount = 0;
    this.#readByIdAndNameIdArgument = undefined;
    this.#readByIdAndNameNameArgument = undefined;
    this.#createInvokeCount = 0;
    this.#createAccountArgument = undefined;
    this.#updateByIdAndNameInvokeCount = 0;
    this.#updateByIdAndNameIdArgument = undefined;
    this.#updateByIdAndNameNameArgument = undefined;
    this.#updateByIdAndNameAccountArgument = undefined;
    this.#deleteByIdAndNameInvokeCount = 0;
    this.#deleteByIdAndNameIdArgument = undefined;
    this.#deleteByIdAndNameNameArgument = undefined;
  }

  set readByNameReturnValue(readByNameReturnValue) {
    this.#readByNameReturnValue = readByNameReturnValue;
  }

  get readByNameInvokeCount() {
    return this.#readByNameInvokeCount;
  }

  get readByNameNameArgument() {
    return this.#readByNameNameArgument;
  }

  set readByIdAndNameReturnValue(readByIdAndNameReturnValue) {
    this.#readByIdAndNameReturnValue = readByIdAndNameReturnValue;
  }

  get readByIdAndNameInvokeCount() {
    return this.#readByIdAndNameInvokeCount;
  }

  get readByIdAndNameIdArgument() {
    return this.#readByIdAndNameIdArgument;
  }

  get readByIdAndNameNameArgument() {
    return this.#readByIdAndNameNameArgument;
  }

  set createReturnValue(createReturnValue) {
    this.#createReturnValue = createReturnValue;
  }

  get createInvokeCount() {
    return this.#createInvokeCount;
  }

  get createAccountArgument() {
    return this.#createAccountArgument;
  }

  set updateByIdAndNameReturnValue(updateByIdAndNameReturnValue) {
    this.#updateByIdAndNameReturnValue = updateByIdAndNameReturnValue;
  }

  get updateByIdAndNameInvokeCount() {
    return this.#updateByIdAndNameInvokeCount;
  }

  get updateByIdAndNameIdArgument() {
    return this.#updateByIdAndNameIdArgument;
  }

  get updateByIdAndNameNameArgument() {
    return this.#updateByIdAndNameNameArgument;
  }

  get updateByIdAndNameAccountArgument() {
    return this.#updateByIdAndNameAccountArgument;
  }

  set deleteByIdAndNameReturnValue(deleteByIdAndNameReturnValue) {
    this.#deleteByIdAndNameReturnValue = deleteByIdAndNameReturnValue;
  }

  get deleteByIdAndNameInvokeCount() {
    return this.#deleteByIdAndNameInvokeCount;
  }

  get deleteByIdAndNameIdArgument() {
    return this.#deleteByIdAndNameIdArgument;
  }

  get deleteByIdAndNameNameArgument() {
    return this.#deleteByIdAndNameNameArgument;
  }
}

export {
  AccountRepositorySpy
};
